### DISCUSSION TITLE
功能讨论

### DISCUSSION THREAD

**Discussion by wangzhankun** [HARM4Y/FOKS-TROT#32] [Ideas]
## 手动加解密的部分
我在想，既然是手动加解密的功能，那就应该不考虑文件扩展名、文件是否处于机密文件夹下面。只要用户有加密需求，就应当给用户进行加密。如果是机密文件夹下的机密拓展名的文件的话，似乎也没有手动加解密的需求。

因此关于手动加解密的部分，我想把它改成无论任何文件都能加密的形式。

**Comment by wangzhankun** [comment]
这个BUG是这样的：
`CommPort.c->PocReentryToEncrypt->PocReentryToGetStreamContext->PocFindOrCreateStreamContext`，由于是要手动加密的文档，不存在`StreamContext`，而在`PocReentryToGetStreamContext`调用`PocFindOrCreateStreamContext`时没有要求创建`StreamContext`。在`PocReentryToEncrypt`中将status修改为了`POC_IRRELEVENT_FILE_EXTENSION`，因此从外界看来是由于文件拓展名导致的。

**Comment by HARM4Y** [comment]
这里要是添加的话，可以在commport.c中把该文件的扩展名加入到管控的范围内，就是加到那个数组中。 @wangzhankun

**Reply by wangzhankun** [reply]
倒不仅仅是拓展名的问题。还有机密文件夹的设置。在PreCreate里面都会被过滤掉。我现在考虑的是增加一个全局的临时缓冲区，用于临时保存需要进行加解密的文件路径。在过滤的时候进行判断即可。然后当该文件执行成功之后就清空临时缓冲区。

我不建议直接添加机密拓展名和机密文件夹，因为这样会导致本身不应当作为机密拓展名和机密文件夹的部分被添加进来了。

这个实现我现在正在调试，今晚应该可以推送上来。

**Reply by HARM4Y** [reply]
暂时先不用推送，我这里还有bug没修完

**Comment by wangzhankun** [comment]
关于通信那里我想重新设计一下，想设计成透明传输的形式。会对PocUserDll和POC->CommPort.c改动很多，你看同意不？

**Reply by HARM4Y** [reply]
什么是透明传输？

**Reply by wangzhankun** [reply]
类似于计网的IP/TCP协议那种。PocUserDll类似于提供了TCP的协议，具体的涉及业务的协议由用户自己设计。比如PocUserDll中的一些添加删除文件夹之类的，这些都是属于拓展功能，由用户进行设计。PocUserDll仅提供最基本的通信协议。在驱动那边也是如此，采用统一的透明协议进行收包，具体的应用协议可以由第三方自行编写新的C文件。这样做便于拓展。

**Reply by wangzhankun** [reply]
因为现在的设计，PocUserDll跟PocUserPanel是高度耦合的。我是希望PocUserDll能够与PocUserPanel解耦合。这样如果要进行第三方开发的话，也可以直接调用PocUserDll，驱动那边也不需要进行特殊处理。只需要设计与自己 业务相关的数据结构和算法逻辑即可。

**Reply by HARM4Y** [reply]
可以试一下吧，你先提交在自己的仓库里吧，要是可以的话，我从你那里下载以后手动合并。不过写的每一处代码都要测试，之前添加机密文件夹的PocAnyPath2DosPath函数有问题，文件夹无法添加成功。

**Comment by wangjianxm** [comment]
商业模式一般重写FSD的,这种的写法一个另存为就失效了,个人感觉作者和何明的论文没啥区别,密文挪用也可以换成CRT的对称加密模式,不过这个微软自带的Bcrypt里面好像没有
